File: back_end/contracts/services/profile_storage.py
"""
User profile storage: one JSON file per user in S3 (users/{username}.json).
All profile data is serialized/deserialized; no DynamoDB table for users or profiles.
"""
import logging
from decimal import Decimal

from django.contrib.auth import get_user_model

from .user_data_s3 import get_user_data, save_user_data

logger = logging.getLogger(__name__)

# Profile dict keys matching API / UserProfileSerializer
PROFILE_ATTRS = [
    "name",
    "total_contract_value",
    "contract_count",
    "certifications",
    "clearances",
    "naics_codes",
    "industry_tags",
    "work_cities",
    "work_counties",
    "capabilities",
    "agency_experience",
    "size_status",
    "contract_types",
    "created_at",
    "updated_at",
    "uploaded_documents",  # list of { id, title, document, created_at } sustained in profile
]


def _username_for_user_id(user_id):
    """Resolve Django user id to username."""
    try:
        User = get_user_model()
        return User.objects.get(pk=int(user_id)).username
    except Exception:
        return None


def _default_profile_dict(user_id):
    from django.utils import timezone

    now = timezone.now().isoformat()
    return {
        "user_id": user_id,
        "name": "",
        "total_contract_value": "0",
        "contract_count": 0,
        "certifications": [],
        "clearances": [],
        "naics_codes": [],
        "industry_tags": [],
        "work_cities": [],
        "work_counties": [],
        "capabilities": [],
        "agency_experience": [],
        "size_status": [],
        "created_at": now,
        "updated_at": now,
        "uploaded_documents": [],
    }


def _json_to_profile(raw, user_id):
    """Convert JSON/dict (from S3) to profile dict for serializer (id = user_id)."""
    if not raw:
        return None
    total = raw.get("total_contract_value")
    if total is not None and not isinstance(total, str):
        total = str(total)
    return {
        "id": user_id,
        "name": raw.get("name") or "",
        "total_contract_value": total if total is not None else "0",
        "contract_count": int(raw.get("contract_count", 0)),
        "certifications": list(raw.get("certifications") or []),
        "clearances": list(raw.get("clearances") or []),
        "naics_codes": list(raw.get("naics_codes") or []),
        "industry_tags": list(raw.get("industry_tags") or []),
        "work_cities": list(raw.get("work_cities") or []),
        "work_counties": list(raw.get("work_counties") or []),
        "capabilities": list(raw.get("capabilities") or []),
        "agency_experience": list(raw.get("agency_experience") or []),
        "size_status": list(raw.get("size_status") or []),
        "created_at": raw.get("created_at") or "",
        "updated_at": raw.get("updated_at") or "",
        "uploaded_documents": list(raw.get("uploaded_documents") or []),
    }


def profile_dict_to_object(profile_dict):
    """Return an object with attributes so UserProfileSerializer can serialize it."""
    if profile_dict is None:
        return None

    class ProfileObj:
        pass

    o = ProfileObj()
    for k, v in profile_dict.items():
        setattr(o, k, v)
    return o


def _profile_to_json(profile_dict):
    """Profile dict -> JSON-serializable dict (no user_id in stored profile blob)."""
    out = {}
    for k in PROFILE_ATTRS:
        v = profile_dict.get(k)
        if v is None and k in ("total_contract_value", "contract_count"):
            v = "0" if k == "total_contract_value" else 0
        if v is None and k == "uploaded_documents":
            v = []
        if v is not None:
            if k == "total_contract_value" and isinstance(v, (int, float, Decimal)):
                v = str(v)
            if k == "uploaded_documents":
                v = list(v) if isinstance(v, (list, tuple)) else []
            out[k] = v
    return out


def get_profile(user_id):
    """
    Load profile from S3 users/{username}.json. Returns profile dict (id, name, ...)
    or None if not found.
    """
    username = _username_for_user_id(user_id)
    if not username:
        return None
    data = get_user_data(username)
    if not data:
        return None
    raw = data.get("profile")
    if not raw:
        return None
    logger.info("Retrieved profile from S3 for user_id=%s username=%s", user_id, username)
    return _json_to_profile(raw, int(user_id))


def save_profile(profile_dict):
    """
    Save profile to S3 users/{username}.json. Preserves token/token_expires_at in the same file.
    Called whenever profile is saved (PATCH /api/profile/) or after contract create/update/delete
    (refresh_profile_from_contracts), so the JSON is constantly updated.
    """
    user_id = profile_dict.get("user_id")
    if user_id is None:
        return
    username = _username_for_user_id(user_id)
    if not username:
        logger.warning("Cannot save profile: no username for user_id=%s", user_id)
        return
    data = get_user_data(username) or {}
    # Preserve token so profile saves don't wipe Bearer auth
    profile_blob = _profile_to_json(profile_dict)
    data["profile"] = profile_blob
    save_user_data(username, data)
    logger.info("Profile saved to S3 for user_id=%s username=%s", user_id, username)


def get_or_create_profile(user_id):
    """
    Load profile from S3; if missing, create default and save, then return it.
    Returns profile dict.
    Ensures applied_rfp_ids and in_progress_rfp_ids are initialised as empty
    lists so new signups never inherit stale data.
    """
    try:
        logger.info("get_or_create_profile: loading from S3 (user_id=%s)", user_id)
        profile = get_profile(user_id)
        if profile is not None:
            return profile
        logger.info("Creating new profile in S3 for user_id=%s", user_id)
        default = _default_profile_dict(user_id)
        # Save profile and explicitly initialise RFP status fields so new
        # users never see stale in-progress or applied RFPs.
        username = _username_for_user_id(user_id)
        if username:
            data = get_user_data(username) or {}
            data["profile"] = _profile_to_json(default)
            data.setdefault("applied_rfp_ids", [])
            data.setdefault("in_progress_rfp_ids", [])
            save_user_data(username, data)
            logger.info("Profile + empty RFP status saved to S3 for user_id=%s", user_id)
        else:
            save_profile(default)
        return _json_to_profile(_profile_to_json(default), user_id)
    except Exception as e:
        logger.warning("get_or_create_profile failed for user_id=%s: %s", user_id, e)
        default = _default_profile_dict(user_id)
        return _json_to_profile(_profile_to_json(default), user_id)


def refresh_profile_from_contracts(user):
    """
    Recompute profile from user's contracts (stored in user JSON) and save to S3 users/{username}.json.
    Called after every contract create, update, and delete so the user JSON stays in sync.
    """
    from django.utils import timezone

    from .contract_storage import list_contracts_for_profile

    contract_list = list_contracts_for_profile(user.id)
    # Filter out documents that don't belong to this user (S3 key mismatch)
    user_prefix = f"uploads/{user.id}/"
    contract_list = [
        c for c in contract_list
        if not (isinstance(c, dict) and c.get("document_s3_key") and not c["document_s3_key"].startswith(user_prefix))
    ]
    certs = set()
    clearances_set = set()
    naics = set()
    tags = set()
    cities = set()
    counties = set()
    capabilities_set = set()
    agencies = set()
    technology_stack = set()
    contract_types = set()
    size_statuses = set()
    contractor_names = set()
    total_val = Decimal("0")

    def _get(obj, key, default=None):
        """Get a value from an object or dict."""
        if isinstance(obj, dict):
            return obj.get(key, default)
        return getattr(obj, key, default)

    # Keywords that indicate size/status designations (not certifications)
    _SIZE_STATUS_KEYWORDS = (
        'small business', 'large business', 'sdb', 'wosb', 'edwosb',
        'hubzone', '8(a)', '8a', 'sdvosb', 'vosb', 'dbe', 'mbe', 'wbe',
        'minority-owned', 'woman-owned', 'women-owned', 'veteran-owned',
        'service-disabled', 'disadvantaged business', 'sba ', 'small disadvantaged',
    )

    for c in contract_list:
        rc = _get(c, "required_certifications") or []
        rcl = _get(c, "required_clearances") or []
        naics_list = _get(c, "naics_codes") or []
        it = _get(c, "industry_tags") or []
        jc = _get(c, "jurisdiction_city")
        jcy = _get(c, "jurisdiction_county")
        wl = _get(c, "work_locations") or []
        ia = _get(c, "issuing_agency")
        wd = _get(c, "work_description")
        val = _get(c, "contract_value_estimate")
        tech = _get(c, "technology_stack") or []
        scope_kw = _get(c, "scope_keywords") or []
        ct = _get(c, "contract_type")
        cn = _get(c, "contractor_name")

        if cn and str(cn).strip():
            contractor_names.add(str(cn).strip())
        # Reclassify size/status designations that may appear in certifications
        for cert_item in (rc or []):
            if cert_item and str(cert_item).strip():
                cert_lower = str(cert_item).lower().strip()
                if any(kw in cert_lower for kw in _SIZE_STATUS_KEYWORDS):
                    size_statuses.add(str(cert_item).strip())
                else:
                    certs.add(str(cert_item).strip())
        clearances_set.update(rcl or [])
        naics.update(naics_list or [])
        tags.update(it or [])
        if jc:
            cities.add(jc)
        if jcy:
            counties.add(jcy)
        cities.update(wl or [])
        if ia:
            agencies.add(ia)
        if wd and str(wd).strip():
            capabilities_set.add(str(wd).strip())
        # Aggregate technology stack from extracted proposals
        for t in tech:
            if t and str(t).strip():
                technology_stack.add(str(t).strip())
        # Aggregate scope keywords as additional capabilities
        for kw in scope_kw:
            if kw and str(kw).strip():
                capabilities_set.add(str(kw).strip())
        # Aggregate contract types
        if ct and str(ct).strip():
            contract_types.add(str(ct).strip())
        # Aggregate size/status designations
        ss = _get(c, "size_status") or []
        if isinstance(ss, list):
            for s in ss:
                if s and str(s).strip():
                    size_statuses.add(str(s).strip())
        elif ss and str(ss).strip():
            size_statuses.add(str(ss).strip())
        try:
            total_val += Decimal(str(val or "0").replace(",", "").replace("$", ""))
        except Exception:
            pass
    now = timezone.now().isoformat()
    profile_dict = get_profile(user.id) or _default_profile_dict(user.id)
    # Keep full contract list in uploaded_documents (source of truth in user JSON; no DynamoDB)
    uploaded_documents = []
    for c in contract_list:
        if isinstance(c, dict):
            uploaded_documents.append(c)
        else:
            cid = getattr(c, "id", None) or getattr(c, "contract_id", None)
            title = getattr(c, "title", None) or ""
            doc = getattr(c, "document", None) or ""
            created = getattr(c, "created_at", None) or ""
            uploaded_documents.append({"id": cid, "title": title, "document": doc, "created_at": created})
    existing_name = profile_dict.get("name") or ""
    if not existing_name and contractor_names:
        profile_dict["name"] = sorted(contractor_names)[0]

    profile_dict.update(
        {
            "user_id": user.id,
            "certifications": list(certs),
            "clearances": list(clearances_set),
            "naics_codes": list(naics),
            "industry_tags": list(tags),
            "work_cities": list(cities),
            "work_counties": list(counties),
            "capabilities": list(capabilities_set),
            "agency_experience": list(agencies),
            "size_status": list(size_statuses),
            "contract_count": len(contract_list),
            "total_contract_value": str(total_val),
            "updated_at": now,
            "uploaded_documents": uploaded_documents,
        }
    )
    save_profile(profile_dict)
    return profile_dict

File: front_end/src/lib/profile-storage.ts
/**
 * User profile storage and aggregation.
 * Port of back_end/contracts/services/profile_storage.py.
 */
import { getUserData, saveUserData, type UserProfile, type UserData } from "./user-data";
import { listContracts } from "./contract-storage";

const PROFILE_ATTRS = [
  "name",
  "total_contract_value",
  "contract_count",
  "certifications",
  "clearances",
  "naics_codes",
  "industry_tags",
  "work_cities",
  "work_counties",
  "capabilities",
  "agency_experience",
  "size_status",
  "contract_types",
  "created_at",
  "updated_at",
  "uploaded_documents",
] as const;

function defaultProfile(): UserProfile {
  const now = new Date().toISOString();
  return {
    name: "",
    total_contract_value: "0",
    contract_count: 0,
    certifications: [],
    clearances: [],
    naics_codes: [],
    industry_tags: [],
    work_cities: [],
    work_counties: [],
    capabilities: [],
    agency_experience: [],
    size_status: [],
    contract_types: [],
    created_at: now,
    updated_at: now,
    uploaded_documents: [],
  };
}

function jsonToProfile(raw: Record<string, unknown>): UserProfile {
  if (!raw) return defaultProfile();
  let total = raw.total_contract_value;
  if (total != null && typeof total !== "string") total = String(total);

  return {
    name: (raw.name as string) || "",
    total_contract_value: (total as string) ?? "0",
    contract_count: Number(raw.contract_count || 0),
    certifications: (raw.certifications as string[]) || [],
    clearances: (raw.clearances as string[]) || [],
    naics_codes: (raw.naics_codes as string[]) || [],
    industry_tags: (raw.industry_tags as string[]) || [],
    work_cities: (raw.work_cities as string[]) || [],
    work_counties: (raw.work_counties as string[]) || [],
    capabilities: (raw.capabilities as string[]) || [],
    agency_experience: (raw.agency_experience as string[]) || [],
    size_status: (raw.size_status as string[]) || [],
    contract_types: (raw.contract_types as string[]) || [],
    created_at: (raw.created_at as string) || "",
    updated_at: (raw.updated_at as string) || "",
    uploaded_documents: (raw.uploaded_documents as any[]) || [],
  };
}

function profileToJson(profile: UserProfile): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const k of PROFILE_ATTRS) {
    let v = (profile as unknown as Record<string, unknown>)[k];
    if (v == null && k === "total_contract_value") v = "0";
    if (v == null && k === "contract_count") v = 0;
    if (v == null && k === "uploaded_documents") v = [];
    if (k === "total_contract_value" && typeof v === "number") v = String(v);
    if (v != null) out[k] = v;
  }
  return out;
}

// ── Public API ──

export async function getProfile(username: string): Promise<UserProfile | null> {
  const data = await getUserData(username);
  if (!data?.profile) return null;
  return jsonToProfile(data.profile as unknown as Record<string, unknown>);
}

export async function saveProfile(
  username: string,
  profile: UserProfile
): Promise<void> {
  const data = (await getUserData(username)) || ({} as UserData);
  data.profile = profileToJson(profile) as unknown as UserProfile;
  await saveUserData(username, data);
}

export async function getOrCreateProfile(username: string): Promise<UserProfile> {
  try {
    const existing = await getProfile(username);
    if (existing) return existing;

    const profile = defaultProfile();
    const data = (await getUserData(username)) || ({} as UserData);
    data.profile = profileToJson(profile) as unknown as UserProfile;
    if (!data.applied_rfp_ids) data.applied_rfp_ids = [];
    if (!data.in_progress_rfp_ids) data.in_progress_rfp_ids = [];
    await saveUserData(username, data);
    return profile;
  } catch (err) {
    console.warn(`getOrCreateProfile failed for username=${username}:`, err);
    return defaultProfile();
  }
}

// Keywords that indicate size/status designations (not certifications)
const SIZE_STATUS_KEYWORDS = [
  "small business",
  "large business",
  "sdb",
  "wosb",
  "edwosb",
  "hubzone",
  "8(a)",
  "8a",
  "sdvosb",
  "vosb",
  "dbe",
  "mbe",
  "wbe",
  "minority-owned",
  "woman-owned",
  "women-owned",
  "veteran-owned",
  "service-disabled",
  "disadvantaged business",
  "sba ",
  "small disadvantaged",
];

/**
 * Recompute profile from all user contracts and save to S3.
 * Port of back_end/contracts/services/profile_storage.py:refresh_profile_from_contracts
 */
export async function refreshProfileFromContracts(
  username: string
): Promise<UserProfile> {
  const contractList = await listContracts(username);

  const certs = new Set<string>();
  const clearancesSet = new Set<string>();
  const naics = new Set<string>();
  const tags = new Set<string>();
  const cities = new Set<string>();
  const counties = new Set<string>();
  const capabilitiesSet = new Set<string>();
  const agencies = new Set<string>();
  const contractTypes = new Set<string>();
  const sizeStatuses = new Set<string>();
  const contractorNames = new Set<string>();
  let totalVal = 0;

  for (const c of contractList) {
    const cn = c.contractor_name;
    if (cn?.trim()) contractorNames.add(cn.trim());

    // Reclassify size/status designations from certifications
    for (const certItem of c.required_certifications || []) {
      if (!certItem?.trim()) continue;
      const certLower = certItem.toLowerCase().trim();
      if (SIZE_STATUS_KEYWORDS.some((kw) => certLower.includes(kw))) {
        sizeStatuses.add(certItem.trim());
      } else {
        certs.add(certItem.trim());
      }
    }

    for (const cl of c.required_clearances || []) {
      if (cl?.trim()) clearancesSet.add(cl.trim());
    }
    for (const n of c.naics_codes || []) {
      if (n?.trim()) naics.add(n.trim());
    }
    for (const t of c.industry_tags || []) {
      if (t?.trim()) tags.add(t.trim());
    }
    if (c.jurisdiction_city) cities.add(c.jurisdiction_city);
    if (c.jurisdiction_county) counties.add(c.jurisdiction_county);
    for (const wl of c.work_locations || []) {
      if (wl?.trim()) cities.add(wl.trim());
    }
    if (c.issuing_agency) agencies.add(c.issuing_agency);
    if (c.work_description?.trim()) capabilitiesSet.add(c.work_description.trim());

    // Technology stack
    for (const t of c.technology_stack || []) {
      if (t?.trim()) capabilitiesSet.add(t.trim());
    }
    // Scope keywords
    for (const kw of c.scope_keywords || []) {
      if (kw?.trim()) capabilitiesSet.add(kw.trim());
    }
    // Contract types
    const ct: unknown = c.contract_type;
    if (Array.isArray(ct)) {
      for (const t of ct) {
        if (typeof t === "string" && t.trim()) contractTypes.add(t.trim());
      }
    } else if (typeof ct === "string" && ct.trim()) {
      contractTypes.add(ct.trim());
    }

    // Size/status from dedicated field
    const ss: unknown = c.size_status || [];
    if (Array.isArray(ss)) {
      for (const s of ss) {
        if (typeof s === "string" && s.trim()) sizeStatuses.add(s.trim());
      }
    } else if (typeof ss === "string" && ss.trim()) {
      sizeStatuses.add(ss.trim());
    }

    // Contract value
    try {
      const valStr = (c.contract_value_estimate || "0")
        .replace(/,/g, "")
        .replace(/\$/g, "");
      totalVal += parseFloat(valStr) || 0;
    } catch {
      // skip invalid values
    }
  }

  const now = new Date().toISOString();
  const existing = (await getProfile(username)) || defaultProfile();

  // Keep full contract list in uploaded_documents
  const uploadedDocuments = contractList.map((c) => ({ ...c }));

  // Auto-set name from contractor names if empty
  if (!existing.name && contractorNames.size > 0) {
    existing.name = [...contractorNames].sort()[0];
  }

  const updated: UserProfile = {
    ...existing,
    certifications: [...certs],
    clearances: [...clearancesSet],
    naics_codes: [...naics],
    industry_tags: [...tags],
    work_cities: [...cities],
    work_counties: [...counties],
    capabilities: [...capabilitiesSet],
    agency_experience: [...agencies],
    size_status: [...sizeStatuses],
    contract_types: [...contractTypes],
    contract_count: contractList.length,
    total_contract_value: String(totalVal),
    updated_at: now,
    uploaded_documents: uploadedDocuments,
  };

  await saveProfile(username, updated);
  return updated;
}
